src/mrinufft/trajectories/densities.py

```diff
@@ -0,0 +1,134 @@
+    rng = nr.default_rng()
+
+    shape = density.shape
+    nb_dims = len(shape)
+    max_nb_samples = np.prod(shape)
+
+    if nb_samples > max_nb_samples:
+        raise ValueError("`nb_samples` must be lower than the size of `density`.")
+
+    if method == "random":
+        choices = rng.choice(
+            np.arange(max_nb_samples),
+            size=nb_samples,
+            p=density.flatten(),
+            replace=False,
+        )
+        locations = np.indices(shape).reshape((nb_dims, -1))[:, choices]
+        locations = locations.T
+        locations = 2 * KMAX * locations / np.max(locations) - KMAX
+    elif method == "lloyd":
+        kmeans = (
+            KMeans(n_clusters=nb_samples)
+            if nb_dims <= 2
+            else BisectingKMeans(n_clusters=nb_samples)
+        )
+        kmeans.fit(
+            np.indices(density.shape).reshape((nb_dims, -1)).T,
+            sample_weight=density.flatten(),
+        )
+        locations = kmeans.cluster_centers_ - np.array(density.shape) / 2
+        locations = KMAX * locations / np.max(np.abs(locations))
+    else:
+        raise ValueError(f"Unknown sampling method {method}.")
+    return locations
+
+
+def create_cutoff_decay_density(shape, cutoff, decay, resolution=None):
+    shape = np.array(shape)
+    nb_dims = len(shape)
+
+    if not resolution:
+        resolution = np.ones(nb_dims)
+
+    distances = np.indices(shape).astype(float)
+    for i in range(nb_dims):
+        distances[i] = distances[i] + 0.5 - shape[i] / 2
+        distances[i] = distances[i] / shape[i] * resolution[i]
+    distances = nl.norm(distances, axis=0)
+
+    cutoff = cutoff * np.max(distances) if cutoff else np.min(distances)
+    density = np.ones(shape)
+    decay_mask = np.where(distances > cutoff, True, False)
+
+    density[decay_mask] = (cutoff / distances[decay_mask]) ** decay
+    density = density / np.sum(density)
+    return density
+
+
+def create_polynomial_density(shape, decay, resolution=None):
+    return create_cutoff_decay_density(
+        shape, cutoff=0, decay=decay, resolution=resolution
+    )
+
+
+def create_energy_density(dataset):
+    nb_dims = len(dataset.shape) - 1
+    axes = range(1, nb_dims + 1)
+
+    kspace = nf.fftshift(nf.fftn(nf.fftshift(dataset, axes=axes), axes=axes), axes=axes)
+    energy = np.mean(np.abs(kspace), axis=0)
+    density = energy / np.sum(energy)
+    return density
+
+
+def create_chauffert_density(shape, wavelet_basis, nb_wavelet_scales, verbose=False):
+    nb_dims = len(shape)
+    indices = np.indices(shape).reshape((nb_dims, -1)).T
+
+    density = np.zeros(shape)
+    unit_vector = np.zeros(shape)
+    for ids in tqdm(indices, disable=not verbose):
+        ids = tuple(ids)
+        unit_vector[ids] = 1
+        fourier_vector = nf.ifftn(unit_vector)
+        coeffs = pw.wavedecn(
+            fourier_vector, wavelet=wavelet_basis, level=nb_wavelet_scales
+        )
```

Comment:
You seem to be doing \Psi F*, the paper describes F* \psi right?

Reply:
I based this on the original code from Nicolas:
- https://github.com/philouc/mri_acq_recon_tutorial/blob/master/matlab/TSP_Markov/tools/compute_optimal_distrib.m
- https://github.com/philouc/mri_acq_recon_tutorial/blob/master/matlab/TSP_Markov/tools/perform_wavortho_transfAnis.m

I checked that the code had an identical behavior, and it is the case except for one major difference: he used hardcoded wavelets banks and I use `pywavelets` instead, and their values seem to differ.